xhydro/modelling/calibration.py

```diff
@@ -0,0 +1,322 @@
+        "model_config" object containing hydrological modelling data required,
+        low and high parameter bounds, as well as an objective function.
+        Depending on the objective function, either spotpy or hydroeval will
+        compute the value, since some functions are found only in one package.
+
+        Notes
+        -----
+        Accepted obj_func values:
+            - "abs_bias" : Absolute value of the "bias" metric
+            - "abs_pbias": Absolute value of the "pbias" metric
+            - "abs_volume_error" : Absolute value of the volume_error metric
+            - "agreement_index": Index of agreement
+            - "bias" : Bias metric
+            - "correlation_coeff": Correlation coefficient
+            - "kge" : Kling Gupta Efficiency metric (2009 version)
+            - "kge_mod" : Kling Gupta Efficiency metric (2012 version)
+            - "mae": Mean Absolute Error metric
+            - "mare": Mean Absolute Relative Error metric
+            - "mse" : Mean Square Error metric
+            - "nse": Nash-Sutcliffe Efficiency metric
+            - "pbias" : Percent bias (relative bias)
+            - "r2" : r-squared, i.e. square of correlation_coeff.
+            - "rmse" : Root Mean Square Error
+            - "rrmse" : Relative Root Mean Square Error (RMSE-to-mean ratio)
+            - "rsr" : Ratio of RMSE to standard deviation.
+            - "volume_error": Total volume error over the period.
+        """
+        # Gather the model_config dictionary and obj_func string, and other
+        # optional arguments.
+        self.model_config = model_config
+        self.obj_func = obj_func
+        self.mask = mask
+        self.transform = transform
+        self.epsilon = epsilon
+        self.take_negative = take_negative
+
+        # Create the sampler for each parameter based on the bounds
+        self.parameters = [
+            Uniform("param" + str(i), bounds_low[i], bounds_high[i])
+            for i in range(0, len(bounds_high))
+        ]
+
+    def simulation(self, x):
+        """Simulation function for spotpy.
+
+        This is where the optimizer generates a parameter set from within the
+        given bounds and generates the simulation results. We add the parameter
+        "x" that is generated by spotpy to the model_config object at the
+        reserved keyword "parameters".
+        """
+        self.model_config.update({"parameters": x})
+
+        # Run the model and return Qsim, with model_config containing the
+        # tested parameter set.
+        return hydrological_model_selector(self.model_config)
+
+    def evaluation(self):
+        """Evaluation function for spotpy.
+
+        Here is where we get the observed streamflow and make it available to
+        compare the simulation and compute an objective function. It has to be
+        in the Qobs keyword, although with some small changes
+        model_config['Qobs'] could be a string to a file. Probably more
```

Comment:
One issue that I could foresee here is that `Qobs` is currently defined as a numpy array, while (for Hydrotel at least) streamflow data is usually stored as NetCDF files, which would be read with `xarray` instead.

Reply:
J'ai updaté le code pour qu'il check si le input est un xarray. S'il l'est, on lit les données en array et le code continue.